oauth2_http/javatests/com/google/auth/oauth2/functional/FTServiceAccountCredentialsTest.java

```diff
@@ -0,0 +1,191 @@
+  public void AudienceSetNoScopeTest() throws Exception {
+    final GoogleCredentials credentials = GoogleCredentials.getApplicationDefault();
+
+    IdTokenCredentials tokenCredential =
+        IdTokenCredentials.newBuilder()
+            .setIdTokenProvider((IdTokenProvider) credentials)
+            .setTargetAudience(cloudTasksUrl)
+            .build();
+
+    assertNull(tokenCredential.getIdToken());
+    tokenCredential.refresh();
+    IdToken idToken = tokenCredential.getIdToken();
+    assertNotNull(idToken);
+    assertTrue(idToken.getExpirationTime().getTime() > System.currentTimeMillis());
+    JsonWebSignature jws =
+        JsonWebSignature.parse(GsonFactory.getDefaultInstance(), idToken.getTokenValue());
+    assertEquals(cloudTasksUrl, jws.getPayload().get("aud"));
+    assertEquals("https://accounts.google.com", jws.getPayload().get("iss"));
+  }
+
+  @Test
+  public void ScopeSetNoAudienceStorageTest() throws Exception {
+    HttpResponse response = executeRequestWithCredentialsWithScope(storageUrl, cloudPlatformScope);
+    assertEquals(200, response.getStatusCode());
+  }
+
+  @Test
+  public void ScopeSetNoAudienceComputeTest() throws Exception {
+
+    HttpResponse response = executeRequestWithCredentialsWithScope(computeUrl, cloudPlatformScope);
+    assertEquals(200, response.getStatusCode());
+  }
+
+  @Test
+  public void ScopeSetNoAudienceBigQueryTest() throws Exception {
+    HttpResponse response = executeRequestWithCredentialsWithScope(bigQueryUrl, cloudPlatformScope);
+    assertEquals(200, response.getStatusCode());
+  }
+
+  @Test
+  public void ScopeSetNoAudienceOnePlatformTest() throws Exception {
+    HttpResponse response =
+        executeRequestWithCredentialsWithScope(cloudTasksUrl, cloudPlatformScope);
+    assertEquals(200, response.getStatusCode());
+  }
+
+  @Test
+  public void WrongScopeComputeTest() throws Exception {
+    executeRequestWrongScope(computeUrl);
+  }
+
+  @Test
+  public void WrongScopeStorageTest() throws Exception {
+    executeRequestWrongScope(storageUrl);
+  }
+
+  @Test
+  public void WrongScopeBigQueryTest() throws Exception {
+    executeRequestWrongScope(bigQueryUrl);
+  }
+
+  @Test
+  public void WrongScopeOnePlatformTest() throws Exception {
+    executeRequestWrongScope(cloudTasksUrl);
+  }
+
+  private void executeRequestWrongScope(String serviceUri)
+      throws FileNotFoundException, IOException {
+    String expectedMessage = "403 Forbidden";
+
+    try {
+      executeRequestWithCredentialsWithScope(
+          serviceUri, "https://www.googleapis.com/auth/adexchange.buyer");
+      fail(String.format("Should throw exception '%s'", expectedMessage));
+    } catch (IOException expected) {
+      assertTrue(expected.getMessage().contains(expectedMessage));
+    }
+  }
+
+  private HttpResponse executeRequestWithCredentialsWithoutScope(String serviceUrl)
+      throws IOException {
+    final GoogleCredentials credentials = GoogleCredentials.getApplicationDefault();
```

Comment:
What type of credential is it? Does Java run these functional tests in kokoro?

Reply:
it is serviceAccountCredential. Yes, it is run in Kokoro like integration tests. There is some common setup for yoshi java projects.